packages/rest/src/__tests__/unit/request-context.unit.ts

```diff
@@ -0,0 +1,72 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {aRestServerConfig, givenRequestContext} from '../helpers';
+
+describe('RequestContext', () => {
+  describe('requestedProtocol', () => {
+    it('defaults to "http"', () => {
+      const ctx = givenRequestContext();
+      expect(ctx.requestedProtocol).to.equal('http');
+    });
+
+    it('honors "x-forwarded-proto" header', () => {
+      const ctx = givenRequestContext(aRestServerConfig(), {
+        url: '/',
+        headers: {'x-forwarded-proto': 'https'},
+      });
+      expect(ctx.requestedProtocol).to.equal('https');
+    });
+
+    it('honors protocol provided by Express request', () => {
+      const ctx = givenRequestContext(aRestServerConfig());
+
+      // tslint:disable-next-line:no-any
+      (ctx.request.connection as any).encrypted = true;
+      expect(ctx.request.protocol).to.equal('https');
+
+      expect(ctx.requestedProtocol).to.equal('https');
+    });
+  });
+
+  describe('basePath', () => {
+    it('defaults to an empty string', () => {
+      const ctx = givenRequestContext(aRestServerConfig());
+      expect(ctx.basePath).to.equal('');
+    });
+
+    it('honors baseUrl when mounted on a sub-path', () => {
+      const ctx = givenRequestContext(aRestServerConfig());
+      // emulate `expressApp.mount('/api', lb4app)
+      ctx.request.baseUrl = '/api';
+      expect(ctx.basePath).to.equal('/api');
+    });
+  });
+
+  describe('requestedBaseUrl', () => {
+    it('defaults to data from the HTTP connection', () => {
+      const ctx = givenRequestContext(
+        aRestServerConfig({
+          host: undefined,
+          port: 0,
+        }),
+      );
+      expect(ctx.requestedBaseUrl).to.equal('http://localhost');
```

Comment:
When port is `0`, doesn't it use a random port? Should't the resulting base URL also contain the port number?

Reply:
Good catch again!

I guess this is a limitation of testing this part in unit-test style. When port is set to `0`, we should be using a random port indeed. I don't know how to mock that up though.

What do you propose?

Reply:
Maybe it does not make sense to unit test this function?

Reply:
Or generate a random port in `givenRequestContext` if port is 0?

Reply:
I wonder why the host is `localhost` instead of `127.0.0.1`.

Reply:
> Maybe it does not make sense to unit test this function?

We should be able to detect breaking changes in the implementation/behavior without having a bug let us know about it.

Reply:
> I wonder why the host is localhost instead of 127.0.0.1

`shot` is making a request to `localhost`. The value of the `host` is whatever the client is requesting, not what was configured.

Reply:
> Or generate a random port in givenRequestContext if port is 0?

Just found out, no matter what `port` value you set, it is always being set to `80` (by `shot`?). Since `80` is the default HTTP port, it is being stripped to return `http://localhost` instead of `http://localhost:80`.

Shouldn't `RequestContext` be tested in the context of an actual `RestServer` instead of a simulator (shot)?

Reply:
> Shouldn't RequestContext be tested in the context of an actual RestServer instead of a simulator (shot)?

Possibly. I wrote the tests at unit-test level because you asked for that:

> Shouldn't we add unit test for this? We want to be able to detect any unexpected change in future.

Reply:
I rewrote all new tests into integration-test style, to verify the real end-to-end behavior.